### ISSUE TITLE
Refresh model zoo

### ISSUE THREAD

**Issue by joost-ve** [FluxML/model-zoo#266]
I started testing all the models in this repo against the flux 0.11.4 release on julia 1.5 (ref https://github.com/FluxML/ML-Coordination-Tracker/issues/9). In this issue i will collect all encoutered problems and try to list all other open issues for each model. The tests are ran against my [fork](https://github.com/joost-ve/model-zoo) which contains the updated manifests and hotfixes needed to actually run some of the models.

**all**
- [ ] Parameters.@with_kw vs Base.@kwdef, see https://github.com/mauro3/Parameters.jl/issues/89. None of the models seem to depend on features not provided by the Base implementation.
- [ ] BSON.jl seems not really well maintained and several models will not be able to use it on julia 1.6 unless this [issue](https://github.com/JuliaIO/BSON.jl/issues/80) is fixed.

**vision/cdcgan_mnist**:
- runs without issues

**vision/cifar10**:
- [ ] dependency problems when updateing to Flux 0.11.4
- [ ] depends on Metalhead dataset instead of MLDatasets

**vison/cppn**:
- [ ] no GPU support

**vison/dcgan_mnist**:
- [ ] fails, probably this [issue](https://github.com/FluxML/Flux.jl/issues/1350)
- [ ] hparams.latent_dim not available in generator constructor [(hotfix)](https://github.com/joost-ve/model-zoo/commit/eb7fd41221bd6650cc5d06d5352d79a68f9e0e6f)
- [ ] uses deprecated logitbinarycrossentropy [(hotfix)](https://github.com/joost-ve/model-zoo/commit/515a81fd5eb988850da5511c5526cca082271e0d)
- [ ] fails on julia 1.6-beta1 because of BSON.jl issue

**vison/lenet_mnist**:
- [ ] depends on CUDAapi instead of CUDA
- [ ] fails on julia 1.6-beta1 because of BSON.jl issue [(hotfix)](https://github.com/joost-ve/model-zoo/commit/ff733cea0f850c487156b89ff2910000a6edc86f)

**vison/mnist/mlp**:
- runs without issues

**vison/mnist/conv**:
- [ ] depends on Flux.Data.MNIST instead of MLDatasets

**vison/mnist/autoencoder**:
- [ ] depends on Flux.Data.MNIST instead of MLDatasets

**vison/vae_mnist**:
- [ ] logitbinarycrossentropy.(ŷ, y) fails and is deprecated [(hotfix)](https://github.com/joost-ve/model-zoo/commit/c4a273e116ba2180dbfbbc8eedc8524b6cd9dbed)
- [ ] depends on CUDAapi instead of CUDA  [(hotfix)](https://github.com/joost-ve/model-zoo/commit/c4a273e116ba2180dbfbbc8eedc8524b6cd9dbed)
- [ ] resulting images are all gray
- [ ] encoder is not moved to GPU before saving resulting in errors in plotting [(hotfix)](https://github.com/joost-ve/model-zoo/commit/da719406482eb2e763099250c3c17ee6e6d2c88b)
- [ ] scalar operations on GPU

**Comment by DhairyaLGandhi** [comment]
Can we get PRs for the hotfixes?

**Comment by CarloLucibello** [comment]
I suggest we also rationalize the examples. We could remove
- vision/mnist/conv
- vision/mnist/autoencoder
- vison/cppn

as redundant or not adding much value (in the last case)

**Comment by CarloLucibello** [comment]
Also we could rename 
- vision/cifar10 ->  vision/vgg_cifar10
- vision/mnist/mlp ->  vision/mlp_mnist

**Comment by darsnack** [comment]
Just for reference, what's the approach here? Are you following the steps in https://github.com/FluxML/ML-Coordination-Tracker/issues/9 ? So, is this to track hot fixes to get things running, and later we'll refresh the files to use the latest packages in the ecosystem?

**Comment by joost-ve** [comment]
Yes, steps as mentioned in https://github.com/FluxML/ML-Coordination-Tracker/issues/9. This issue is to keep track of the test results, but I plan to note down other issues found along the way.

For the models that fail to run I do update the dependencies before writing a fix. And if a package is the source of the problem I intend to replace it with the latest ecosystem package one. Take for example the cifar10 model, this one can not be used with the newest flux release due to the dependency on Metalhead for it's dataset, with MLDatasets it can actually be tested on flux 0.11.4.

**Comment by DhairyaLGandhi** [comment]
I wouldn't remove the models, those are some of the better maintained models, compared to how the some others have bloated argument handling taking away from the core of the solved problem. Removing those as redundant seems like the better approach no?

**Comment by joost-ve** [comment]
I do agree, some of the "simpler" models could actually serve as a great example of a "my first model". But I do think we should have that discussion in the coordination tracker issue, and keep this issue about the quick fixes to get the models running again.

**Comment by joostveenema** [comment]
The other/flux-next model is a bit problematic. This one is rather old and seems to explain a 'new' api for the optimisers, but this does not work (anymore). It probably needs a full rewrite, but than it wouldn't really add something to the official Flux docs. I would suggest to just remove it from the zoo or does anyone think it still has a place here?

**Comment by DhairyaLGandhi** [comment]
For now, we can leave it be, it will be part of the Optimisers.jl release. It's not expected to be a guarantee but maybe adding a readme saying so would be alright

**Comment by joostveenema** [comment]
Are you sure? It does not seems to match with the api from the Optimisers.jl package (it uses a step! function)..

**Comment by DhairyaLGandhi** [comment]
It's from a flux pr that we plan to merge, but it's a breaking release so may not do it immediately.

**Comment by DhairyaLGandhi** [comment]
Can we add an item in the tracker about adding guiding texts with the models that talk about:
A. The model, and the problem being solved
B. The specific features in flux that we are trying to communicate (custom structs, loss, training loop, model construction etc)

**Comment by DhairyaLGandhi** [comment]
cc @darsnack ^

**Comment by darsnack** [comment]
Sure, did you mean zoo tutorials?

**Comment by DhairyaLGandhi** [comment]
Basically having text pointing users to how to use certain Flux features, and also guiding them about different problems beings solved (recurrent nets, basic features and regression, custom train loops etc.)

If we can use the Literate.jl format, we would be in good stead to automate moving them to the site.

**Comment by CarloLucibello** [comment]
We can remove contrib/games/cartpole and pendulum, curated Flux's based examples are available in https://github.com/JuliaReinforcementLearning/ReinforcementLearningZoo.jl

**Comment by DhairyaLGandhi** [comment]
Those are not there for the RL, they are there to show what the dP equivalent would look like, so I would update then with the literature and maybe add them as ci examples, because I'm sure recent changes have to support this.

**Comment by joostveenema** [comment]
Now that all the models are tested I created a new issue (https://github.com/FluxML/model-zoo/issues/280) to start giving some direction to the next steps and address the last comments made here.

**Comment by DhairyaLGandhi** [comment]
We still need to update Metalhead I take it.

**Comment by joostveenema** [comment]
My comment wasn't meant to skip steps. I'm just trying to prevent doing work now that will be flushed down the drain later in the process.

**Comment by DhairyaLGandhi** [comment]
Sure, i was ensuring that that was a pending item that's blocking some of the updates.

**Comment by DhairyaLGandhi** [comment]
Metalhead doesn't bound flux, also I have released patches to both. We might want to consolidate the environments and only have one at the top level.

It's hard to keep track otherwise

**Comment by joostveenema** [comment]
Do you mean a top level project and manifest file for all the models/tutorials?

That could become problematic when one of the tutorials depends on a package that is not compatible with the newest Flux/CUDA. It could hold back updating of all the models until that external package has been made compatible. Transformers.jl is an example, if one tutorial would depend on it everthing will be held back.

**Comment by DhairyaLGandhi** [comment]
That's not been a big issue since we only need to activate the env, and updating the global env is usually painless, whereas individual ones usually end up becoming inconsistent very quickly and have the same issue of holding back updates. With the added annoyance of having to update many of them for every minor change in a dependency.

**Comment by DhairyaLGandhi** [comment]
We used to have this pattern before and that's where this learning comes from..

**Comment by darsnack** [comment]
I agree that a single Project.toml and Manifest.toml is better. Allowing certain tutorials to fall out of compatibility with the latest Flux is not what we want. Every tutorial should be running on the latest Flux or the latest Flux/the tutorial should be fixed.

The flow should be "Release Flux" --> "Update zoo" --> "Release zoo."

**Comment by darsnack** [comment]
My only concern is that a user needs to install lots of packages to run one specific example. Maybe it is better to have a script that automates bumping Flux for all the tutorials.

**Comment by DhairyaLGandhi** [comment]
> My only concern is that a user needs to install lots of packages

Yeah mine too, but that's still not too bad. For instance, things like CUDA would dominate the space/ network usage for most users and adding packages doesn't take too long these days.

**Comment by DhairyaLGandhi** [comment]
What's the last column for?

**Comment by sambitdash** [comment]
I really like the MNIST models. So please do not remove if you can. The reason being, they are pry the only model that can run reasonably on a CPU. Other vision models need GPUs for good accuracy. I will set up this expection from the MNIST example. 

1. The mlp one is a quick example to showcase ANN. 
2. The CNN should show the benefits of using a lot less parameters. Hence, the current one may need some fixing and also take into consideration the data loaders etc. 
3. I think the lautoencoder should preferably use a CNN based encoder architecture.

**Comment by ToucheSir** [comment]
Ideally any model in the zoo that can run on GPU should also converge just as well on CPU. I don't think the MNIST models are going anywhere either!

As for the autoencoder example, we should probably have _both_ dense and conv examples. That may be redundant with the VAE models though.
